### DISCUSSION TITLE
Disable tokenization

### DISCUSSION THREAD

**Discussion by trailerparkdev** [meilisearch/product#422] [Feedback & Feature Proposal]
Hi @curquiza.

This is to follow suit on your request to open a  discussion about our use case.

In a  nutshell, we index strings that are first tokenized  (stemming, stopwords filtering, custom-processing and customer-based enhancements). The text (out of which the strings are extracted) come from repositories that can be multilingual. We "tokenize" accordingly.

These tokenized strings are the only searchable attributes in our documents. They are ingested in a dedicated index which is our main index among several (the others are ancillary).

Of course, when we implement a search, we tokenize the query text the very same way. That's how matching is done.

Based on what you told me, and from tests I ran, I don't see that the mandatory tokenization on your side is a problem. 

My question about disabling tokenization on your side was pure engineer concern about resources, since in this case it *may* not be of much help (unless I'm wrong).

Do you need anything more?

Have a nice day.

---

EDIT by @curquiza 
Following: https://github.com/meilisearch/meilisearch/issues/2257

**Comment by curquiza** [comment]
Hello @trailerparkdev, thanks a lot for sharing this!

@gmourier will come back to you if he needs more information :)

**Comment by gmourier** [comment]
Hi @trailerparkdev 👋

I'm super curious to know why you don't let Meilisearch do the work about the tokenization instead of it doing it on your side!

Thanks

**Comment by trailerparkdev** [comment]
Hi @gmourier.

I'm not really at liberty to discuss that, but i guess you can imagine that depending on the use case, some data need heavy pre-processing...

...and that some of this pre-processing may include use-case specific tokenization...

...which in turn may include stemming / lemmatizing / POS tagging and other kinds of NLP preparation, especially on multi-lingual datasets.

Another angle to plead for disabling tokenization is that some use cases require complete control on tokenization (ie on its results). Which calls for in-house tokenization. And since this process can be complex, and because classic tools of the trade like spaCy or nltk are known to yield imperfect results, even on specifically trained models, the only way you can commit on results for a customer usecase is to be able to master the process end to end. In such situations, MS' tokenization on top of this work may somehow break it. 

Again, I perfectly understand the rationale behind your tokenization process, in that it allows many of MS' great features like typo tolerance, etc. It all makes perfect sense and gives developers a great experience if the usecase matches. If it diverges, that's where we need options to disable auto features like tokenization, in order to be able to take charge of it ourselves, and still benefit of the simplicity and efficiency of the rest of MS. 

Have a nice day.

**Reply by gmourier** [reply]
Thanks @trailerparkdev !

**Comment by noe** [comment]
I subscribe to this need. In my case, I keep language-specific indexes. This would allow me to implement tokenization of whatever language without needing to modify Meilisearch itself.

Also, the built-in tokenizations in Meilisearch may not be appropriate for some cases. For instance, for Chinese, Jieba can be configured with custom dictionaries. I rely on this in the rest of my system, but Meilisearch Jieba-based tokenization uses the default dictionaries.

**Reply by curquiza** [reply]
Hello @noe 
We have opened a [dedicated discussion](https://github.com/meilisearch/product/discussions/503) on Chinese tokenization. Indeed, we have issues with the Chinese language due to Jieba, but we don't know what could be the best alternative since no one is Chinese on our team. Feel free to give your opinion on it, it would be really helpful 🙏

**Comment by tobiasnitsche** [comment]
Hi,
i also have a problem, where i would like to disable tokenization for a specific field.

The reason is because i have several fields (like vehicle license plate, article numbers, customer numbers etc..), which should not be tokenized.

My detailed use case can be found in this thread: https://github.com/meilisearch/meilisearch/discussions/3380

**Reply by gmourier** [reply]
Hello @tobiasnitsche 👋

Interesting take! We have a related [discussion](https://github.com/meilisearch/product/discussions/160) that seems related to your need.

Why do you completely want to disable the tokenization on a field? Is it a faster way to avoid selecting specific separators to keep at tokenization?

Thanks!

**Reply by tobiasnitsche** [reply]
Hi @gmourier 

i have described my use case in this topic: https://github.com/meilisearch/meilisearch/discussions/3380

Basically I have fields (next to fields like a article description etc.) which are more or less "exact" values and could have for exampe an "-" or an "/". 

For exampe an article-number can be "A-123-BCDF" or "A/2123-2323". For those specific fields I would love to have the tokenizer disabled (similar to typo correction)

So the user can enter both start an aproximate search (like on article description) or an exact search (like on article nr) in the same index / request.

Does it help you? Im bad at describing ;-)

**Reply by tobiasnitsche** [reply]
@gmourier  is there any chance of implementing one day? I know its not the highest priority for you all, but im just regulary checking the roadmap and couldnt find it ;-) https://roadmap.meilisearch.com/

**Reply by macraig** [reply]
Hi @tobiasnitsche ! We've just released v1.4.0 that adds the capability of customizing tokenization by adding or removing separators to the separator and non-separator lists of tokens. You can read more about it in our docs:

- [Non-separator tokens](https://www.meilisearch.com/docs/reference/api/settings#non-separator-tokens)
- [Separator tokens](https://www.meilisearch.com/docs/reference/api/settings#separator-tokens)

Would this solve your use case?

**Reply by tobiasnitsche** [reply]
Hi @macraig 
no it does not. I described my problem a few times.

I have a database  , lets say users. The user has a first name, last name and an email address. 

**i need to update or disable the tokenizations on field base**


Use case:

An e-mail address like "test-email_address@whatever.com" should *not* be tokenized, the rest should work normal.

Please be honest, as i need to plan: 


Is the any chance that it will ever be integrated or do i need to switch the search engine?  We have this problem on multiple fields like article numbers, project numbers, customer numbers, license plate....

**Reply by curquiza** [reply]
Hello @tobiasnitsche 

I'm not sure I understand why you cannot search you email address and you have to disable the tokenization
I just made a test by adding a document as `[{ "id": 1, "mail": "test-email_address@whatever.com"}]` and I can search for "email",  or "test", or "add", or "address", or "whatever" and the document is always retrieved.

Can you give the exact example of what you try to do, what you get, and what you would like to have instead?

**Reply by tobiasnitsche** [reply]
Hi @curquiza 

i described my problem in here:

https://github.com/meilisearch/meilisearch/discussions/3380

**Reply by tobiasnitsche** [reply]
Its not only about e-mail addresses. Also like license plate. Imagine our users are searching for a license plate like "HH-XY 234". People are entering it like "HH-XY", and they expect to find only plates starting with "HH-XY" and not plates having "XY". 

To summarize: I need disabling tokenization on field base.

**Reply by curquiza** [reply]
Which version are you using?

Because I just made a test with these documents
```
[
  { "id": 1, "text": "HH-XY" },
  { "id": 2, "text": "XY" }
]
```

And then search for "HH-XY", and I only get doc 1, not doc 2

Can you give another example so that I can really understand your pain point?

**Reply by limness** [reply]
@trailerparkdev We have encountered a similar problem in our team. Although I also think that for some fields disabling tokenization would be useful.

A partial solution is to remove the necessary delimiters so that the search is relatively "hard":
https://www.meilisearch.com/docs/reference/api/settings#non-separator-tokens

**Reply by ManyTheFish** [reply]
Hello @tobiasnitsche @limness,

With @macraig we are thinking of adding a new setting. This will allow the user to precise the list of attributes that should be considered as identifiers, the values of these fields will not be tokenized and will not allow any typo.

At search time, nothing changes in the API; the end-user will be able to search as before, but the query will be processed twice:
1) the query is tokenized, and the search engine retrieves the documents based on the searchable attributes
2) the query is taken row, and the search engine retrieves the documents based on the identifier attributes

There are limits on forbidding tokenization, for instance, using `space`s instead of `TAB`s will not retrieve the document containing an identifier with `TAB`s.

Could this setting help in your use case?
 
 Thanks!

**Reply by tobiasnitsche** [reply]
Hi @ManyTheFish and @macraig 

The setting you have proposed would perfectly solve my Problem!

To my example, i will try my best to give you a better example. 

But maybe try this:

Articles Nr

ABC-50-744-YX
YX-50-ABC

As article numbers are sensitive

ABC-50 should only return the First



(But the table might have another column "Description" which should work with tokenization.

Many of users have huge article databases, but know the article numbers of their most used articles

**Reply by tobiasnitsche** [reply]
I will give you details of our Version hopefully by tomorrow.

**Reply by tobiasnitsche** [reply]
Hi @curquiza 

Here is an example.

My targetfor this attribute is to have only articles having number "DEF-" at the beginning. For the description field and title field it should work as normal with tokenization. Hope this helps / underlines my problem ;-)

![grafik](https://github.com/meilisearch/product/assets/43043989/881b45b2-7426-46e2-9e9a-9eaeed3d3426)

**Reply by ManyTheFish** [reply]
Hello @tobiasnitsche, 
Is it possible for you to set your `Articles Nr` as filterable attributes, use the [facet search](https://www.meilisearch.com/docs/reference/api/facet_search#perform-a-facet-search) to retrieve the value and apply the filter?
If it's not possible, is the "1 search for all fields" the only blocker, or is there something else?

**Reply by tobiasnitsche** [reply]
Thank you for your reply @ManyTheFish and sorry for being that late.

Thats actually the only blocker yes, I would prefer and "1 search for all field" solution. 

We have never tried face search yet. Will look into that, as soon as we find time.

**Comment by tobiasnitsche** [comment]
Hi Guillaume ,

 

exactly. For the typo correction i disabled it already on the fields like "article-nr", "license-plate" etc...

Having the same option to disable tokenize on attributes would be my awesome wish :-)

 
 

Gesendet: Dienstag, 31. Januar 2023 um 12:21 Uhr
Von: "Guillaume Mourier" ***@***.***>
An: "meilisearch/product" ***@***.***>
Cc: "tobiasnitsche" ***@***.***>, "Mention" ***@***.***>
Betreff: Re: [meilisearch/product] Disable tokenization (Discussion #422)


 

Thanks! FYI, you can already disable the typo correction for the field containing the article number. You want to use disableOnAttributes. https://blog.meilisearch.com/typo-tolerance/

So you prefer disabling the whole tokenization of a field instead of keeping it in a sparse way by customizing the separators for a given field?

**Comment by paulRbr** [comment]
> Having the same option to disable tokenize on attributes would be my awesome wish :-)

Adding a comment here to emphasize this sentence from @tobiasnitsche. We (at bump.sh) would also love to be able to customize the “non separator tokens” on one single attribute :pray:
